feat: supporting failOnThreshold

dylanmorley opened the pull request:
As per #137, this adds support for a new parameter named `failOnThreshold`, which works as an additional control for detecting violations. 

An example use-case might be when you want to report on **all** detections to your project, but you only want to fail the build on critical thresholds. Previously, this wasn't possible as the severityThreshold was both the detection and failure threshold, whereas now we can split these out with a configuration such as

```
- task: SnykSecurityScan@1
  inputs:
    serviceConnectionEndpoint: 'mySnykToken'
    testType: 'app'
    failOnIssues: true
    failOnThreshold: 'critical'
    severityThreshold: 'low'
    monitorWhen: 'always'
```

As part of this change
- Additional tests around the new feature
- Split out validation logic from index->parseInputArgs into new method validate on TaskArgs - supports easier testing

Fixes #137

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/snyk/snyk-azure-pipelines-task?pullRequest=143) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/snyk/snyk-azure-pipelines-task?pullRequest=143) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/snyk/snyk-azure-pipelines-task?pullRequest=143) it.</sub>

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -1,5 +1,9 @@
 import * as tl from 'azure-pipelines-task-lib';
-
+import {
+  isNotValidThreshold,
```
Since you moved validation logic here, `isNotValidThreshold` should be also a part of TaskArgs class now.

michelkaporin commented on snykTask/src/task-lib.ts:
```diff
@@ -3,6 +3,7 @@ import * as tr from 'azure-pipelines-task-lib/toolrunner';
 import * as tl from 'azure-pipelines-task-lib/task';
 import stream = require('stream');
 import * as fs from 'fs';
+const fsPromises = require('fs').promises;
```
Let's stick with import syntax, as it's in other imports:
`import * as fsPromises from 'fs/promises'`.

michelkaporin commented on README.md:
```diff
@@ -27,6 +27,7 @@ This extension requires that Node.js and npm be installed on the build agent. Th
 | dockerfilePath            | The path to the Dockerfile corresponding to the `dockerImageName`                                                                                                                                                   | yes, if container image test | none          | string                                                                            |
 | targetFile                | Applicable to application type tests ony. The path to the manifest file to be used by Snyk. Should only be provided if non-standard.                                                                                | no                           | none          | string                                                                            |
 | severityThreshold         | The severity-threshold to use when testing. By default, issues of all severity types will be found.                                                                                                                 | no                           | "low"         | string: "low" or "medium" or "high" or "critical"                                 |
+| failOnThreshold           | The severity-threshold to use when determining if the build should fail. Combine with `failOnIssues` to get fine grained control over reporting and build failure behaviour. For example, with `failOnIssues` set to true and `failOnThreshold` to `critical`, all issues would be reported on but only critical issues would cause a build failure                               ...
```
Could you please mention that `severityThreshold` parameter is responsible for reporting? We should avoid people expecting that  `failOnThreshold` is also responsible for reporting.

michelkaporin commented on snykTask/src/__tests__/test-task-args.ts:
```diff
@@ -103,6 +104,38 @@ describe('TaskArgs.setMonitorWhen', () => {
   });
 });
 
+describe('TaskArgs.severitySettings', () => {
```
`TaskArgs.validate` naming would be consistent with other group of tests, where they're grouped by a function name (e.g. `TaskArgs.setMonitorWhen`).

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -10,7 +14,7 @@ class TaskArgs {
 
   severityThreshold: string | undefined = '';
-
+  failOnThreshold: string | undefined = 'low';
```
You don't need `undefined` in type definition because you provide default value for the property.

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -72,6 +76,32 @@ class TaskArgs {
+
+  public validate() {
+    if (this.failOnThreshold){
+      if (isNotValidThreshold(this.failOnThreshold!)) {
```
Non-null assertion is unnecessary here and should generally be avoided, relying on if-statements instead.

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -72,6 +76,32 @@ class TaskArgs {
+    }
+
+    if (this.severityThreshold) {
+      if (isNotValidThreshold(this.severityThreshold!)) {
```
Non-null assertion is unnecessary.

michelkaporin commented on snykTask/src/index.ts:
```diff
@@ -402,7 +399,11 @@ async function run() {
       taskArgs.failOnIssues
     ) {
-      throw new SnykError(snykTestResult.message);
+      let failureThreshold:string = taskArgs.failOnThreshold!
```
You won't need non-null assertion operator here once you remove `undefined` declaration in the property definition (see my other comment).

michelkaporin reviewed (commented):
Like that you've encapsulated validation logic in the TaskArgs class 👍

dylanmorley commented:
Thanks - addressed issues + I've tried to be more explicit about what the parameters do in the README, updated example etc.

michelkaporin commented:
@dylanmorley we run integration test as part of the deployment pipeline and I see `fs/promises` import is breaking task execution.
<img width="655" alt="Screenshot 2022-06-09 at 16 11 29" src="https://user-images.githubusercontent.com/2239563/172867824-eca4c101-435d-44d4-aa58-6bdc389a8b6c.png">
 
Have you tested it or does it work in your environment? I am not really sure why it cannot find the import since Node.js version is v16.15.1 on our build agent. How about we convert `readFile()` to synchronous `fs.readFileSync()` to fix it? We're not getting much benefit from async pattern here anyway, and it's going to be consistent with other code, since they do sync FS operations.

dylanmorley commented:
Ran a few more tests against different targets (dotnet / maven), few issues to resolve - last commits :

- Switched over to sync FS methods
- Handle cases when scanned target has multiple projects (e.g. dotnet solution with +1 project files) and therefore creates a JSON file with the vulnerabilities reported as an array. 
    - added tests around these output types 
- More defensive check for json file existence in failure threshold logic
    - default to 'detected' if json file not found, maintains backwards compat
    - better logging in this method  

![image](https://user-images.githubusercontent.com/5038454/173028784-d6907cdb-2606-48ae-9bde-075b3df22753.png)

- Included an example VSCode launch config in contributing, a quick start to see what env variables need setting for debugging

NB : Was having problems with VSCode launch around the json report generation, resulting in, 

```
TypeError: require(...).internalModuleReadJSON is not a function
    at internalModuleReadJSON (internal/modules/package_json_reader.js:4:68)
    at Object.read (internal/modules/package_json_reader.js:21:34)
```
Running from command line fine, so I think it's around version VSCode launches with - didn't get to root cause of this, just used command line from windows with env vars set e.g. 

![image](https://user-images.githubusercontent.com/5038454/173043861-5072f3e8-c384-46d9-9e12-7525fbf18f17.png)

michelkaporin commented on .github/CONTRIBUTING.md:
````diff
@@ -9,6 +9,37 @@ Unit tests can be run via `npm run test:unit` command.
+A number of environment variable are required for debugging, here's an example launch config for `VSCode` that sets mandatory parameters such as `AGENT_TEMPDIRECTORY`, `INPUT_failOnIssues` and `INPUT_authToken`
+
+```
+{
````
I've tried to run this configuration and it didn't work for me. I was getting exit code 9 from `snykTestToolRunner.exec()`. Did it work for you without any additional setup?

I've noticed you are using Windows, whereas I am on Mac. If that works on Windows, would be good to note it here.

dylanmorley replied:
Exit code 9 from that call is the error I mentioned regarding `internalModuleReadJSON is not a function` . 

Full root cause is https://blog.sethcorker.com/question/getting-to-the-bottom-of-mix-tailwind-exited-with-1/, summarized as '[Packaged apps crash with zero messaging when NODE_OPTIONS is set](https://github.com/vercel/pkg/issues/1517)' and VSCode setting NODE_OPTIONS on launch. 

Setting to `null` seems to do the trick, will include that in next change

michelkaporin commented on .github/CONTRIBUTING.md:
```diff
@@ -48,6 +79,8 @@ To release a major you need to add `BREAKING CHANGE: ` to the start of the body
 
 Ensure that your code adheres to the included `.eslintrc` config by running `npm run test:checks`.
 
+Check your code is formatted by running `prettier` e.g. `npx prettier --check snykTask/**/*.ts` and auto-fix any violations by performing `npx prettier --write snykTask/**/*.ts`
```
There're dedicated scripts for this already in package.json.

- `npm run test:checks` runs prettier verification (--check), as stated in the previous sentence
- `npm run format` runs formatter (--write)

michelkaporin commented on snykTask/src/__tests__/task-lib.test.ts:
```diff
@@ -45,6 +46,47 @@ test('getOptionsToExecuteSnyk builds IExecOptions like we need it', () => {
+  expect(itemsFound).toBe(true);
+});
+
+test('ignores vulnerabilities lower than high threshold', () => {
```
I would avoid using "ignore" word because it has a special meaning at Snyk. Let's rename to "not finds vulnerabilities..." or "not matches vulnerabilities..."

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -72,6 +72,33 @@ class TaskArgs {
+
+    if (this.severityThreshold) {
+      if (this.isNotValidThreshold(this.severityThreshold)) {
+        const errorMsg = `If set, severityThreshold must be '${Severity.CRITICAL}' or '${Severity.HIGH}' or '${Severity.MEDIUM}' or '${Severity.LOW}' (case insensitive). If not set, the default is 'low'.`;
```
Let's replace 'low' with enum value '${Severity.LOW}' at the end :)

michelkaporin commented on snykTask/src/task-args.ts:
```diff
@@ -72,6 +72,33 @@ class TaskArgs {
+  public validate() {
+    if (this.failOnThreshold) {
+      if (this.isNotValidThreshold(this.failOnThreshold)) {
+        const errorMsg = `If set, failOnThreshold must be '${Severity.CRITICAL}' or '${Severity.HIGH}' or '${Severity.MEDIUM}' or '${Severity.LOW}' (case insensitive). If not set, the default is 'low'.`;
```
Let's replace 'low' with enum value '${Severity.LOW}' at the end :)

michelkaporin reviewed (commented):
Sorry for taking long to review this, had no chance to look into it. These are great points you've uncovered. I've left the comments. Also noticed `snykTask/src/__tests__/task-lib-severity-threshold.test.ts` was deleted. Any reason why?

I'll be OOO next week, thus it's @bastiandoetsch or @asaf92 who'll look into the PR, if they get time. Otherwise, I'll do the last review once back on 27th June. Please communicate to support, if there's urgency. Thanks!

dylan-asos commented:
Thanks - will get those addressed. Re deleting task-lib-severity-threshold.test.ts, we've moved the function `isNotValidThreshold` that file was testing to be part of the TaskArgs, so moved the tests from that file so they're now part of TaskArgs.validate

michelkaporin reviewed (approved):
Looks good to me ✅ Thank you and apologise for making you wait that long. I'll notify you when it's publicly released.

dylanmorley commented:
Thanks @michelkaporin let me know if any issues, happy to support
